ui/src/app/components/demographic-survey-v2.tsx

```diff
@@ -0,0 +1,794 @@
+import * as React from 'react';
+import { useEffect, useState } from 'react';
+import * as fp from 'lodash/fp';
+import validate from 'validate.js';
+
+import {
+  EducationV2,
+  EthnicCategory,
+  GenderIdentityV2,
+  Profile,
+  SexAtBirthV2,
+  SexualOrientationV2,
+} from 'generated/fetch';
+
+import { FlexColumn, FlexRow } from 'app/components/flex';
+import { SemiBoldHeader } from 'app/components/headers';
+import { withProfileErrorModal } from 'app/components/with-error-modal';
+import colors, { colorWithWhiteness } from 'app/styles/colors';
+import { reactStyles } from 'app/utils';
+
+import { Divider } from './divider';
+import { CheckBox, NumberInput, TextInput } from './inputs';
+import { MultipleChoiceQuestion } from './multiple-choice-question';
+import { AoU } from './text-wrappers';
+import { WithSpinnerOverlayProps } from './with-spinner-overlay';
+import { YesNoOptionalQuestion } from './yes-no-optional-question';
+
+const maxYear = new Date().getFullYear();
+const minYear = maxYear - 125;
+const styles = reactStyles({
+  question: { fontWeight: 'bold', color: colors.primary },
+  answer: { margin: '0.0rem 0.25rem', color: colors.primary },
+});
+
+const validateDemographicSurvey = (demographicSurvey) => {
+  validate.validators.nullBoolean = (v) =>
+    v === true || v === false || v === null
+      ? undefined
+      : 'value must be selected';
+
+  // Not well documented, but in Validate.JS, if your message is prefixed with a caret,
+  // the field name will not be included with the message:
+  // https://validatejs.org/docs/validate.html#section-47
+  // https://github.com/ansman/validate.js/issues/68
+  const yearOfBirth = demographicSurvey.yearOfBirthPreferNot
+    ? {}
+    : {
+        yearOfBirth: {
+          presence: {
+            allowEmpty: false,
+            message:
+              '^You must either fill in your year of birth or check the corresponding "Prefer not to answer" checkbox',
+          },
+          numericality: {
+            onlyInteger: true,
+            greaterThanOrEqualTo: minYear,
+            lessThanOrEqualTo: maxYear,
+          },
+        },
+      };
+  const validationCheck = {
+    ethnicCategories: { presence: { allowEmpty: false } },
+    genderIdentities: { presence: { allowEmpty: false } },
```

Comment:
missing sexualOrientations
also need checks for otherTexts: if category includes OTHER, then the corresponding otherText must not be empty

Reply:
Good call. I suspect that how I implemented validation for the Other text fields can be done more efficiently, so I added a todo comment to look into improving it later.